File: python/508. Most Frequent Subtree Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findFrequentTreeSum(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        mp = collections.defaultdict(int)
        self.max_val = 0
        def dfs(root):
            if not root: return 0
            l = dfs(root.left)
            r = dfs(root.right)
            s = l + r + root.val
            mp[s] += 1
            self.max_val = max(self.max_val, mp[s])
            return s

        dfs(root)
        res = []
        return [k for k, v in mp.iteritems() if v==self.max_val]

File: go/src/leetcode/508/508. Most Frequent Subtree Sum.go
/**
@author ZhengHao Lou
Date    2022/06/19
*/
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findFrequentTreeSum(root *TreeNode) []int {
	var mx int
	var counter = make(map[int]int)
	var ans []int

	var dfs func(node *TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		s := node.Val + dfs(node.Left) + dfs(node.Right)
		counter[s]++
		return s
	}

	dfs(root)
	for s, c := range counter {
		if c > mx {
			mx = c
			ans = []int{s}
		} else if c == mx {
			ans = append(ans, s)
		}
	}
	return ans
}
